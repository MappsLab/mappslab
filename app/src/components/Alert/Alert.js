// @flow
import * as React from 'react'
import styled, { css } from 'styled-components'
import Pane from 'Components/Pane'
import { Button } from 'Components/Buttons'
import type { PromisedAlertConfig } from './AlertProvider'

const Background = styled.div`
	${({ theme }) => css`
		z-index: ${theme.layout.z.alert};
		position: fixed;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
		display: flex;
		justify-content: center;
		align-items: center;
		background-color: rgba(0, 0, 0, 0.15);
	`}
`

/**
 * Alert
 */

type AlertProps = {
	visible: boolean,
	alert?: PromisedAlertConfig,
}
const defaultButtons = [
	{
		label: 'OK',
		level: 'primary',
		returnValue: true,
	},
]

const Alert = ({ visible, alert }: AlertProps) => {
	if (!visible || !alert) return null
	const { title, message, buttons } = alert
	console.log(alert)
	const alertButtons = buttons || defaultButtons
	return (
		<Background data-testid="alert">
			<Pane size="small" title={title}>
				<p>{message}</p>
				{alertButtons.map(({ label, ...buttonConfig }) => (
					<Button key={label} {...buttonConfig}>
						{label}
					</Button>
				))}
			</Pane>
		</Background>
	)
}

Alert.defaultProps = {
	alert: undefined,
}

export default Alert
