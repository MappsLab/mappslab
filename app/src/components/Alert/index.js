// @flow

export { AlertProvider, AlertConsumer } from './AlertProvider'
export { default as Alert } from './Alert'
