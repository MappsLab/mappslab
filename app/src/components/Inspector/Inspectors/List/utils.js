// @flow
import type { ClassroomType, UserType, MapType } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<NodeArray> | NodeArray

export type ListItemHandler = (ClassroomType | UserType | MapType) => void

export type ListItem = {
	node: ClassroomType | UserType | MapType,
	key: string,
	onClick: () => void,
}

// const nodeToInspectorItem = (item: ClassroomType | UserType | MapType): InspectorItem => ({
// 	uid: item.uid,
// 	type: item.__typename,
// 	title: item.__typename === 'User' ? item.name : item.title,
// })

export const nodeToListItem = (node: ClassroomType | UserType | MapType, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
