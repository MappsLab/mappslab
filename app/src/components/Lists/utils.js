// @flow
import type { Node, Variables } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

// type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<void> | void

export type ListItemHandler = (Node) => void

export type ListItem = {
	node: Node,
	onClick: () => void,
}

export type ListOfTypeBaseProps<Type> = {
	title: string,
	items: Array<Type>,
	viewerCanAdd: boolean,
	update: (Type) => void,
	onItemClick: ListItemHandler,
}

export type ListOfTypeProps<Type> = ListOfTypeBaseProps<Type> & {
	searchQuery: (Variables) => Promise<void>,
	searchResults: Array<Type>,
}

export const nodeToListItem = (node: Node, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
