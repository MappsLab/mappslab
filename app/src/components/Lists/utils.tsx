import { NodeType } from '../../types-ts'

export type SearchForList = (arg0: any) => Promise<void> | void

export type ListItemHandler<T extends NodeType> = (
	item: T,
) => void | Promise<void>

export type CreateNewFn = (str: string) => void | Promise<void>

export interface ListItem {
	key: string
	node: NodeType
	info: Array<string>
	onClick: () => void | Promise<void>
}

export interface ListOfTypeProps<Type extends NodeType> {
	title: string
	items?: Type[]
	viewerCanAdd: boolean
	update?: (node: NodeType) => void | Promise<void>
	onItemClick?: ListItemHandler<Type>
	onSearchResultClick?: ListItemHandler<Type>
	create?: (args: any) => Promise<void> | void
	remove?: (args: any) => Promise<void> | void
}

export const nodeToListItem = <T extends NodeType>(
	node: T,
	handler: ListItemHandler<T>,
): ListItem => {
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
