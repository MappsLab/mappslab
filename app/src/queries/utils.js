// @flow
import * as R from 'ramda'
import type { LoadingState } from 'Types/GraphQL'

type Node = {
	cursor: string,
	node: any,
}

type withEdges = {
	edges: Array<Node>,
	pageInfo?: {
		hasNextPage: boolean,
		hasPrevPage: boolean,
		cursor: string,
	},
}

const splitEdges = (obj: withEdges) => ({
	edges: obj.edges.map(({ cursor, node }) => ({ __cursor: cursor, ...node })),
	pageInfo: obj.pageInfo || {},
})

// export const unwindEdges = (obj: any) =>
// 	typeof obj === 'object'
// 		? Object.entries(obj).reduce((acc, [key, value]) => {
// 				if (value && value.edges && Array.isArray(value.edges)) {
// 					// $FlowFixMe
// 					const { edges, pageInfo } = splitEdges(value)
// 					const pageInfoKey = `${key}PageInfo`
// 					return {
// 						[key]: edges.map(unwindEdges),
// 						[pageInfoKey]: pageInfo,
// 						...acc,
// 					}
// 				}
// 				const parsed =
// 					typeof value === 'object'
// 						? // unwind recursively if there are more values
// 						  Array.isArray(value)
// 							? // if it's an array, unwind each item as a new array
// 							  value.map(unwindEdges)
// 							: // otherwise, pass in the object as is
// 							  unwindEdges(value)
// 						: // otherwise, value is OK
// 						  value
// 				return {
// 					[key]: parsed,
// 					...acc,
// 				}
// 		  }, {})
// 		: obj

export const unwindEdges = (o: Object): Object => {
	if (typeof o !== 'object') return o
	return R.pipe(
		R.toPairs,
		// Iterate over the properties and their values
		R.reduce(
			(acc, [key, value]) =>
				value && value.edges && Array.isArray(value.edges)
					? // When the value has an 'edges' property that is an array,
					  R.pipe(
							// Pull out the 'pageInfo' prop and rename it
							R.assoc(`${key.replace(/Connection$/, '')}PageInfo`, value.pageInfo || {}),
							// And pluck out the nodes
							R.assoc(key.replace(/Connection$/, ''), R.pluck('node', value.edges).map(unwindEdges)),
					  )(acc)
					: // otherwise, if it's an object (and not null bc null is an object)
					  value && typeof value === 'object'
						? // unwind it
						  R.assoc(key, Array.isArray(value) ? value.map(unwindEdges) : unwindEdges(value))(acc)
						: // lastly, return the value as is
						  R.assoc(key, value)(acc),
			{},
		),
	)(o)
}

export const getNetworkStatus = (num: number): LoadingState => {
	switch (num) {
		case 1:
			return 'loading'
		case 4:
			return 'refetching'
		case 2:
		case 6:
			return 'passivelyRefetching'
		case 3:
			return 'fetchingMore'
		case 7:
			return 'ready'
		case 8:
			return 'errors'
		default:
			throw new Error(`Status number ${num} is not a valid network status`)
	}
}
