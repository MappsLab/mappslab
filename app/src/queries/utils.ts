import * as R from 'ramda'

type LoadingState =
	| 'loading'
	| 'refetching'
	| 'passivelyRefetching'
	| 'fetchingMore'
	| 'ready'
	| 'errors'

type Node = {
	cursor: string
	node: any
}

type withEdges = {
	edges: Array<Node>
	pageInfo?: {
		hasNextPage: boolean
		hasPrevPage: boolean
		cursor: string
	}
}

const splitEdges = (obj: withEdges) => ({
	edges: obj.edges.map(({ cursor, node }) => ({ __cursor: cursor, ...node })),
	pageInfo: obj.pageInfo || {},
})

export const getNetworkStatus = (num: number | void): LoadingState => {
	switch (num) {
		case undefined:
		case 1:
			return 'loading'
		case 4:
			return 'refetching'
		case 2:
		case 6:
			return 'passivelyRefetching'
		case 3:
			return 'fetchingMore'
		case 7:
			return 'ready'
		case 8:
			return 'errors'
		default:
			throw new Error(`Status number ${num} is not a valid network status`)
	}
}
