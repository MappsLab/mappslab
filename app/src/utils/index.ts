export * from './data'
export * from './listeners'
export * from './maps'
export * from './media'
export * from './statecharts'
export * from './storage'
export * from './url'
export * from './parsing'
