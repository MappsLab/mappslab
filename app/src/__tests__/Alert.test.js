// @flow
/* eslint-disable no-undef */
import * as React from 'react'
import { fireEvent } from 'react-testing-library'
import { render } from 'Jest/utils'
import { AlertProvider, AlertConsumer } from 'Components/Alert'
import type { ContextValues } from 'Components/Alert/AlertProvider'

const AlertWrapper = ({ children }: { children: (ContextValues) => React.Node }) => (
	<AlertProvider>
		<AlertConsumer>{children}</AlertConsumer>
	</AlertProvider>
)

const BasicAlert = () => (
	<AlertWrapper>
		{({ createAlert }: ContextValues) => (
			<button
				type="button"
				data-testid="button"
				onClick={() => {
					createAlert({ message: 'are you sure?', buttons: [{ label: 'YES' }, { label: 'NO', level: 'secondary' }] })
				}}
			>
				click?
			</button>
		)}
	</AlertWrapper>
)

describe('Alert Component', () => {
	it('Mounts Correctly', () => {
		const { container } = render(<BasicAlert />)
		expect(container).toMatchSnapshot()
	})

	it('should render its children', async () => {
		const { getByTestId } = render(<BasicAlert />)
		expect(getByTestId('button')).toHaveTextContent('click?')
	})

	it('should not render an <Alert> by default', async () => {
		const { queryByTestId } = render(<BasicAlert />)
		expect(queryByTestId('alert')).toBeFalsy()
	})

	it('should render an <Alert> when a consumer calls `createAlert', async () => {
		const { queryByTestId, queryByText, getByTestId } = render(<BasicAlert />)
		const button = getByTestId('button')
		expect(queryByTestId('alert')).toBeFalsy()
		fireEvent.click(button)
		expect(queryByTestId('alert')).toBeTruthy()
		expect(queryByText('are you sure?')).toBeTruthy()
		expect(queryByText('YES')).toBeTruthy()
		expect(queryByText('NO')).toBeTruthy()
	})

	it.skip('should supply a "OK" button that returns `true` by default', async () => {
		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})

	it.skip('should display custom buttons when given `actions` config', async () => {
		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})

	it.skip('should return the value specified in the `actions.[n].returns` config', async () => {
		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})

	it.skip('should call the function specified in the `actions.[n].returns` config', async () => {
		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})
})
