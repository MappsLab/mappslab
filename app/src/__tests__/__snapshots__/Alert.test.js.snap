// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component Mounts Correctly 1`] = `
<div>
  <button
    data-testid="button"
    type="button"
  >
    click?
  </button>
</div>
`;
